### PR TITLE
&&: Aruna & Kate

### PR DESCRIPTION
**Pull request by Oh-KPond**
# Scrabble
Congratulations! You're submitting your assignment.

## Comprehension Questions

Feature | Feedback
--- | ---
What are the advantages of using git when collaboratively working on one code base? | It was helpful for creating sandboxes for our code, as well as being able to use our own environments/computers. It helped if we ever wanted to go back to a previous state.
Why do you think we chose to make the `score` method in the `Scoring` class a **class method** instead of an **instance method**? | A class method can be used for all instances instead of just one instance or "player". Since a game has many players, the score should be accessible to all of those players.
Describe an example of a test you wrote for a _nominal case_. | The nominal cases, for example, were those that checked to see if the player responded to play or name.
Describe an example of a test you wrote for an _edge case_. | We did write an edge case for the player play method to make sure that a player could not enter a word that includes special characters.  
Does your code use any methods from the `Enumerable` mixin? If so, where and why was it helpful? | We did use `max_by` and `min_by` to help with the code that decided a winner when there was a tie. It was helpful because it cut down on code that needed to be changed, and made the code base less bulky.
What was one method you and your pair used to debug code? | We used the ruby gem Awesome Print a lot to see what our variable outputs were, and Minitest was great to see where errors were happening.
Is there a specific piece of code you'd like feedback on? | We would like to see feedback on testing and how that can be done better.
What are two discussion points that you and your pair discussed when giving/receiving feedback from each other that you would be willing to share? | When we were thinking about the code, but not pair programming, it was great to use Slack as a way to communicate and give feedback about our ideas. We really thought it was important that we were open with each other throughout the program so that that we could remind ourselves of how we had originally planned to grow.

### CODE REVIEW

File: specs/scoring_spec.rb [hunk 1]
```diff
@@ -10,53 +10,75 @@
...
     it 'returns the only word in a length-1 array' do
+      # Arrange
+      # Act
+      # Assert
+      Scrabble::Scoring.highest_score_from(["dog"]).must_equal "dog"
     end
 
     it 'returns the highest word if there are two words' do
+
+      Scrabble::Scoring.highest_score_from(["dog", "rooster"]).must_equal "rooster"
```
Comment (CheezItMan) [inline specs/scoring_spec.rb:66]: In these tests you should also vary up the order, just to make sure the order of the elements isn't determining the answer. 
```ruby
      Scrabble::Scoring.highest_score_from(["dog", "rooster"]).must_equal "rooster"
     Scrabble::Scoring.highest_score_from(["rooster", "dog"]).must_equal "rooster"
```

Also what about testing a list of words with an invalid word in the mix.

File: lib/scoring.rb [hunk 2]
```diff
@@ -1,9 +1,66 @@
...
 module Scrabble
   class Scoring
     def self.score(word)
+      word = word.downcase
+      letters_array = word.split(//)
+
+      score = 0
+      letters_array.each do |letter|
+        case letter
+        when "a" ,"e","i","o", "u", "l", "n", "r" ,"s","t"
```
Comment (CheezItMan) [inline lib/scoring.rb:12]: This works, but is there a more compact way to do this using a hash?

File: specs/scoring_spec.rb [hunk 3]
```diff
@@ -10,53 +10,75 @@
...
       Scrabble::Scoring.score('').must_be_nil
     end
   end
 
   describe 'highest_score_from' do
     it 'returns nil if no words were passed' do
+      # Arrange
+      # Act
+      # Assert
+      Scrabble::Scoring.highest_score_from([]).must_be_nil
```
Comment (CheezItMan) [inline specs/scoring_spec.rb:54]: Good, now what about an invalid word in the list?

File: lib/player.rb [hunk 4]
```diff
@@ -0,0 +1,60 @@
+require 'awesome_print'
+require_relative '../lib/scoring.rb'
+require 'pry'
+module Scrabble
+  class Player
+    attr_reader :name, :plays, :total
+    attr_accessor :tiles
```
Comment (CheezItMan) [inline lib/player.rb:7]: This should **not** be an accessor.

File: lib/player.rb [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+    def initialize(name)
+      @name = name
+      @plays = Array.new
+      @total = 0
+      @tilebag = Scrabble::Tilebag.new
```
Comment (CheezItMan) [inline lib/player.rb:12]: Each player should **NOT** receive their own TileBag.  Instead use the parameter provided to `draw_tiles`.

File: lib/player.rb [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+      @tiles = @tilebag.draw_tiles(7)
+    end
+
+    def play(word)
+      if word =~ [/[a-zA-Z]+/]
+        return nil
+      end
+
+      @plays << word
```
Comment (CheezItMan) [inline lib/player.rb:21]: This will add the word to `plays` even if they have won previously.

File: lib/player.rb [hunk 7]
```diff
@@ -0,0 +1,60 @@
+require 'awesome_print'
+require_relative '../lib/scoring.rb'
+require 'pry'
+module Scrabble
+  class Player
+    attr_reader :name, :plays, :total
```
Comment (CheezItMan) [inline lib/player.rb:6]: If you don't want the user accessing it, **don't** make `total` an `attr_reader`  You have a `total_score` method, use that instead.

File: lib/player.rb [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+      @plays << word
+      if won?
+        return false
+      else
+        return Scrabble::Scoring.score(word)
+      end
+    end
+
+    def total_score
+      @total = 0
```
Comment (CheezItMan) [inline lib/player.rb:30]: Why make this an instance variable?

File: lib/player.rb [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+
+      @plays.each do |word|
+        @total += Scrabble::Scoring.score(word)
+      end
+
+      return @total
+    end
+
+    def won?
+      @total >= 100 ? (return true) : (return false)
```
Comment (CheezItMan) [inline lib/player.rb:40]: This code assumes that `total_score` has been called **before** `won?` has been called.  Not a good assumption!

Instead just call `total_score` in this method.

File: specs/player_spec.rb [hunk 10]
```diff
@@ -0,0 +1,125 @@
...
+    it "responds to the instance variables " do
+
+      @player.must_respond_to :name
+      @player.name.must_equal "kate"
+    end
+
+    it "Creates a player with an Array of played words" do
+
+      @player.must_respond_to :plays
+      #@player.plays.must_equal ["dog","cat","rooster"]
```
Comment (CheezItMan) [inline specs/player_spec.rb:26]: Why is this commented out?

File: specs/player_spec.rb [hunk 11]
```diff
@@ -0,0 +1,125 @@
...
+
+    it "returns an array of words" do
+
+      @player.plays.must_equal ["dog","cat","rooster"]
+      @player.plays.must_be_kind_of Array
+    end
+
+
+    it "returns false if the player has won" do
+      @player.play("qqqqqqqq")
```
Comment (CheezItMan) [inline specs/player_spec.rb:57]: It should only return false if the player has won prior to playing this word.  This is not the case here.

File: specs/player_spec.rb [hunk 12]
```diff
@@ -0,0 +1,125 @@
...
+      @player.total_score.must_be_kind_of Integer
+      @player.must_respond_to :total_score
+    end
+  end
+
+  describe "#won? method" do
+    it "returns true if total_score greater than or equal to 100" do
+
+      @player.play("qqqq")
+      @player.total_score
```
Comment (CheezItMan) [inline specs/player_spec.rb:88]: `won?` shouldn't fail if the user doesn't call `total_score` first.

File: specs/player_spec.rb [hunk 13]
```diff
@@ -0,0 +1,125 @@
...
+    it "returns the score of a highest_scoring_word" do
+
+      @player.highest_word_score.must_equal 57
+      @player.must_respond_to :highest_word_score
+    end
+  end
+
+  describe "# draw_tiles method" do
+    it "Should be a collection of letters" do
+
```
Comment (CheezItMan) [inline specs/player_spec.rb:112]: You're not testing the `draw_tiles` method!

File: lib/tilebag.rb [hunk 14]
```diff
@@ -0,0 +1,61 @@
...
+
+    def draw_tiles(num)
+      @drawn_tiles = @tiles.sample(num)
+      @tiles.delete(@drawn_tiles)
+
+      return @drawn_tiles
+    end
+
+    def tiles_remaining
+     @tiles.length - @drawn_tiles.length
```
Comment (CheezItMan) [inline lib/tilebag.rb:57]: This isn't working

File: lib/player.rb [hunk 15]
```diff
@@ -0,0 +1,60 @@
...
+      return Scrabble::Scoring.highest_score_from(@plays)
+    end
+
+    def highest_word_score
+      return Scrabble::Scoring.score(highest_scoring_word)
+    end
+
+    def  draw_tiles(tilebag)
+      num = 7 - @tiles.length
+        replace_tiles = Scrabble::Tilebag.draw_tiles(num)
```
Comment (CheezItMan) [inline lib/player.rb:53]: You're calling `draw_tiles` like a class method, not an instance method of `tilebag`

Should be:
```ruby
tilebag.draw_tiles(num)
```

File: specs/tilebag_spec.rb [hunk 16]
```diff
@@ -0,0 +1,46 @@
...
+    it "should include a collection of default tiles" do
+
+    @tilebag.tiles.must_be_kind_of Array
+    @tilebag.tiles.length.must_equal 98
+    end
+  end
+
+  describe '#draw_tiles(num)' do
+    it "returns an array of tiles" do
+      @tilebag.draw_tiles(9)
```
Comment (CheezItMan) [inline specs/tilebag_spec.rb:30]: This test is failing

**Comment by CheezItMan** [comment]
# Scrabble
## What We're Looking For

Feature | Feedback
--- | ---
**General** |
Answered comprehension questions |  Check, I'm glad it sounds like you worked well together. 
Both Teammates contributed to the codebase |  Check
Regular Commits with meaningful commit messages |  Check, good commit messages and regular commits
**Scoring class** |
`score` method |  Check 
Uses appropriate data structure to store the letter score |  There is a more efficient way to do this, consider a hash.
Appropriately handles edge cases |  Check
`highest_score_from` method |  Check
Appropriately handles edge cases |  You need to consider invalid words
Test implementations for `highest_score_from`|  See my in-code notes, you also need to write tests to ensure the order isn't causing a word to win despite another rule.
**Player class** |
Uses an array to keep track of words played | Check
Uses existing code to create `highest_scoring_word` and `highest_word_score` |  Check
Returns `true` or `false` from `won?` method | Check, but see my notes in your code.
Tests edge cases on the `play(word)` |  Some
**TileBag class** |
Uses a data structure to keep the set of default tiles in the bag |  Check, good data structure, but poor choice in how you remove tiles from the bag. 
`draw_tiles` method uses the parameter to draw a set of random tiles |  You are calling `draw_tiles` as if it's a class method and not an instance method.  See my incode notes. 
`tiles_remaining` method returns a count |  **NOT WORKING**
Tests `draw_tiles` edge cases |  You tried, but they're failing.
Summary | You hit most of the learning goals.  You need to look more closely at your `TileBag` class and how tiles are being drawn, you're on the right track, but need more testing and more work on `draw_tiles`

File: lib/scoring.rb [hunk 17]
```diff
@@ -1,9 +1,66 @@
...
+      if array_of_words.length == 1
+        return array_of_words[0]
+      end
+
+      # Create & Populate array of words:
+      winning_word_array = []
+      if array_of_words == []
+        return nil
+      elsif array_of_words.length == 1
+        winning_word_array << array_of_words
```
Comment (CheezItMan) [inline lib/scoring.rb:54]: If you're just going to return 1 word, why push `array_of_words` into `winning_word_array` instead `return array_of_words[0]`
